### PR TITLE
feat(ante): allow doing vesting txs based on whitelist

### PR DESCRIPTION
**Pull request by danwt**
## Description

----

Closes https://github.com/dymensionxyz/rollapp-evm/issues/193

Related https://github.com/dymensionxyz/rollapp-evm/issues/169

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----;

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

**Comment by danwt** [comment]
tested locally 👍

File: app/ante/ante.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+)
+
+type HasPermission = func(ctx sdk.Context, accAddr sdk.AccAddress, perm string) bool
+
+func MustCreateHandler(codec codec.BinaryCodec, txConfig client.TxConfig, maxGasWanted uint64, hasPermission HasPermission, accountKeeper evmtypes.AccountKeeper, stakingKeeper evmosvestingtypes.StakingKeeper, bankKeeper evmtypes.BankKeeper, feeMarketKeeper evmosanteevm.FeeMarketKeeper, evmKeeper evmosanteevm.EVMKeeper, ibcKeeper *ibckeeper.Keeper, distrKeeper anteutils.DistributionKeeper) sdk.AnteHandler {
+	ethOpts := evmosante.HandlerOptions{
+		Cdc:                    codec,
+		AccountKeeper:          accountKeeper,
+		BankKeeper:             bankKeeper,
+		ExtensionOptionChecker: evmostypes.HasDynamicFeeExtensionOption,
```
Comment (mtsitrin) [inline app/ante/ante.go:32]: should add
`/ethermint.types.v1.ExtensionOptionsWeb3Tx` to allowed extensions

File: app/ante/handlers.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+		evmante.NewEthSetUpContextDecorator(options.EvmKeeper),
+		// Check eth effective gas price against the node's minimal-gas-prices config
+		evmante.NewEthMempoolFeeDecorator(options.EvmKeeper),
+		// Check eth effective gas price against the global MinGasPrice
+		evmante.NewEthMinGasPriceDecorator(options.FeeMarketKeeper, options.EvmKeeper),
+		evmante.NewEthValidateBasicDecorator(options.EvmKeeper),
+		evmante.NewEthSigVerificationDecorator(options.EvmKeeper),
+		evmante.NewEthAccountVerificationDecorator(options.AccountKeeper, options.EvmKeeper),
+		evmante.NewCanTransferDecorator(options.EvmKeeper),
+		evmante.NewEthVestingTransactionDecorator(options.AccountKeeper, options.BankKeeper, options.EvmKeeper),
```
Comment (mtsitrin) [inline app/ante/handlers.go:29]: `NewEthVestingTransactionDecorator` not needed as we don't use evmos's vesting module

File: app/ante/handlers.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	)
+}
+
+// Deprecated: NewLegacyCosmosAnteHandlerEip712 creates an AnteHandler to process legacy EIP-712
+// transactions, as defined by the presence of an ExtensionOptionsWeb3Tx extension.
+func newLegacyCosmosAnteHandlerEip712(options HandlerOptions) sdk.AnteHandler {
+	return sdk.ChainAnteDecorators(
+		cosmosDecorators(
+			options,
+			func(c *codectypes.Any) bool { return true },                                                       // allows any extension, this is consistent with the evmos behavior
```
Comment (mtsitrin) [inline app/ante/handlers.go:54]: allow only `/ethermint.types.v1.ExtensionOptionsWeb3Tx` or pass options.ExtensionOptionChecker

File: app/ante/handlers.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+			}, []string{
+				sdk.MsgTypeURL(&vestingtypes.MsgCreateVestingAccount{}),
+				sdk.MsgTypeURL(&vestingtypes.MsgCreatePermanentLockedAccount{}),
+				sdk.MsgTypeURL(&vestingtypes.MsgCreatePeriodicVestingAccount{}),
+			}),
+		ante.NewValidateMemoDecorator(options.AccountKeeper),
+		cosmosante.NewMinGasPriceDecorator(options.FeeMarketKeeper, options.EvmKeeper),
+		ante.NewConsumeGasForTxSizeDecorator(options.AccountKeeper),
+		cosmosante.NewDeductFeeDecorator(options.AccountKeeper, options.BankKeeper, options.DistributionKeeper, options.FeegrantKeeper, options.StakingKeeper, options.TxFeeChecker),
+		cosmosante.NewVestingDelegationDecorator(options.AccountKeeper, options.StakingKeeper, options.Cdc),
```
Comment (mtsitrin) [inline app/ante/handlers.go:93]: not sure what it is. we don't have it on the hub e.g

**Review by mtsitrin — CHANGES_REQUESTED** [review]

File: app/ante/handlers.go (hunk 4 above)
Comment (danwt) [reply app/ante/handlers.go:97]: hmm 
it's from evmos
https://github.com/dymensionxyz/evmos/blob/2dcb56724c6cd1610db205af5ee12d84b608828f/app/ante/handler_options.go#L133

File: app/ante/ante.go (hunk 1 above)
Comment (danwt) [reply app/ante/ante.go:43]: I think it's more consistent with the evmos codebase to not do that

Comment (danwt) [reply app/ante/ante.go:43]: the legacy one does not have extension options checker
https://github.com/dymensionxyz/evmos/blob/2dcb56724c6cd1610db205af5ee12d84b608828f/app/ante/handler_options.go#L119-L171

File: app/ante/handlers.go (hunk 3 above)
Comment (danwt) [reply app/ante/handlers.go]: see https://github.com/dymensionxyz/rollapp-evm/pull/216/files#r1592489840

File: app/ante/handlers.go (hunk 2 above)
Comment (danwt) [reply app/ante/handlers.go]: removed

File: app/ante/handlers.go (hunk 3 above)
Comment (mtsitrin) [reply app/ante/handlers.go]: in evmos they have seperate ChainAnteDecorators
here u reuse `cosmosDecorators`
no reason not to validate the extension in this case

File: app/ante/handlers.go (hunk 4 above)
Comment (mtsitrin) [reply app/ante/handlers.go:95]: no reason to keep it

Comment (danwt) [reply app/ante/handlers.go:95]: ok done 👍

File: app/ante/handlers.go (hunk 3 above)
Comment (danwt) [reply app/ante/handlers.go]: I'm not sure I got you
they don't validate it for the legacy pathway, it's missing in between these lines
https://github.com/dymensionxyz/evmos/blob/2dcb56724c6cd1610db205af5ee12d84b608828f/app/ante/handler_options.go#L152-L153

File: app/ante/ante.go (hunk 1 above)
Comment (danwt) [reply app/ante/ante.go:43]: gonna move convo to https://github.com/dymensionxyz/rollapp-evm/pull/216#discussion_r1592346429

File: app/ante/handlers.go (hunk 3 above)
Comment (danwt) [reply app/ante/handlers.go]: anyways I added a change

**Review by mtsitrin — DISMISSED** [review]
perfect

**Review by omritoptix — APPROVED** [review]

**Pull request merged**